### PR TITLE
fix(auth): add missing attributes, update custom attribute logic

### PR DESCRIPTION
**Pull request by Jordan-Nelson**
*Issue #, if available:*

*Description of changes:*
- Add `sub` and `identifies` attributes
- Update logic for handling custom attributes
- Update sample app

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/auth/amplify_auth_cognito/example/lib/main.dart [hunk 1]
```diff
@@ -54,20 +54,27 @@ class _MyAppState extends State<MyApp> {
     GoRoute(
       path: '/update-user-attribute',
       builder: (BuildContext _, GoRouterState state) =>
-          const UpdateUserAttributeScreen(),
+          UpdateUserAttributeScreen(
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito/example/lib/main.dart:57]: Note: The updates here were just to make testing easier or fix broken routes. They aren't related to the changes in the PR.

File: packages/auth/amplify_auth_cognito/example/lib/screens/confirm_user_attribute.dart [hunk 2]
```diff
@@ -86,7 +86,7 @@ class _ConfirmUserAttributeScreenState
           children: [
             const SizedBox(height: 12),
             TextFormField(
-              initialValue: widget.userAttributeKey.toString(),
+              initialValue: widget.userAttributeKey.key,
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito/example/lib/screens/confirm_user_attribute.dart:89]: Note: The change here is out of scope of this PR, and just to make testing easier. I believe `toString()` use to return `key`.  Since `CognitoUserAttributeKey` is `AWSDebuggable`, that is no longer true.

`userAttributeKey.key` feels a little clunky. I think that was why `toString()` returned the key. Thoughts on calling this `name` rather than `key` in vNext?

File: packages/amplify_core/lib/src/types/auth/attribute/user_attribute_key.dart [hunk 3]
```diff
@@ -21,10 +21,10 @@ import 'package:meta/meta.dart';
 @immutable
 abstract class UserAttributeKey implements Comparable<UserAttributeKey> {
   /// {@macro amplify_core.user_attribute_key}
-  const UserAttributeKey();
+  const UserAttributeKey(this.key);
 
   /// The JSON key for this attribute.
-  String get key;
+  final String key;
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/auth/attribute/user_attribute_key.dart:27]: I kinda prefer the other way so that you can define `key` as a field or a method in subclasses.

File: packages/auth/amplify_auth_cognito/example/lib/screens/confirm_user_attribute.dart (hunk 2 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito/example/lib/screens/confirm_user_attribute.dart:89]: Yep, good catch. 

`name` conflicts with the `CognitoUserAtributeKey.name` which is why I used `key`.

File: packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart [hunk 4]
```diff
@@ -26,35 +26,52 @@ import 'package:meta/meta.dart';
...
+    super.key, {
+    this.readOnly = false,
+    this.isCustom = false,
+  });
 
   /// Creates a custom Cognito attribute.
-  const CognitoUserAttributeKey.custom(this._key)
-      : isCustom = true,
-        readOnly = false;
+  factory CognitoUserAttributeKey.custom(String key) {
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart:36]: Not a huge deal. I'd prefer this being `const` and using the `key` override like before, though.

**Review by dnys1 — APPROVED** [review]
Looks good, thanks!

File: packages/auth/amplify_auth_cognito/example/lib/screens/confirm_user_attribute.dart (hunk 2 above)
Comment (Jordan-Nelson) [reply packages/auth/amplify_auth_cognito/example/lib/screens/confirm_user_attribute.dart:89]: Oh right. i think we have had this conversation lol. I'll think about what else it could be.

File: packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart (hunk 4 above)
Comment (Jordan-Nelson) [reply packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart:36]: I was trying to simplify this, but yeah - maybe the small amount of extra complexity is worth it to make this a const constructor like the others.

Comment (dnys1) [reply packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart]: For sure, yeah it's always a balance

File: packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart [hunk 5]
```diff
@@ -29,32 +29,49 @@ class CognitoUserAttributeKey extends UserAttributeKey
   const CognitoUserAttributeKey._(this._key, {this.readOnly = false})
       : isCustom = false;
 
-  /// Creates a custom Cognito attribute.
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart:32]: Can we add back the doc comment

File: packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart [hunk 6]
```diff
@@ -29,32 +29,49 @@ class CognitoUserAttributeKey extends UserAttributeKey
...
-      orElse: () => CognitoUserAttributeKey.custom(key),
+      orElse: () => CognitoUserAttributeKey._unknown(key),
     );
   }
 
   /// Prefix for custom Cognito attributes.
-  static const customPrefix = 'custom:';
+  static const _customPrefix = 'custom:';
+
+  /// Wether or not a key has starts with the custom prefix that is
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart:62]: sp

**Review by dnys1 — APPROVED** [review]

**Review by haverchuck — APPROVED** [review]

File: packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart [hunk 7]
```diff
@@ -29,32 +29,49 @@ class CognitoUserAttributeKey extends UserAttributeKey
   const CognitoUserAttributeKey._(this._key, {this.readOnly = false})
       : isCustom = false;
 
-  /// Creates a custom Cognito attribute.
   const CognitoUserAttributeKey.custom(this._key)
```
Comment (Jordan-Nelson) [inline packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart:32]:
```suggestion
  /// Creates a custom Cognito attribute.
  const CognitoUserAttributeKey.custom(this._key)
```

File: packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart [hunk 8]
```diff
@@ -34,27 +34,45 @@ class CognitoUserAttributeKey extends UserAttributeKey
...
-      orElse: () => CognitoUserAttributeKey.custom(key),
+      orElse: () => CognitoUserAttributeKey._unknown(key),
     );
   }
 
   /// Prefix for custom Cognito attributes.
-  static const customPrefix = 'custom:';
+  static const _customPrefix = 'custom:';
+
+  /// Wether or not a key has starts with the custom prefix that is
```
Comment (Jordan-Nelson) [inline packages/amplify_core/lib/src/types/auth/attribute/cognito_user_attribute_key.dart:63]:
```suggestion
  /// Whether or not a key starts with the custom prefix that is
```
